### PR TITLE
fix triple quoted new line on type formatting

### LINKED ISSUE #3817: `formatOnType` misbehaves with interpolated triple-quoted multiline strings

### PR DESCRIPTION
**Pull request by vzmerr**
fixes #3817

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala [hunk 1]
```diff
@@ -232,6 +266,16 @@ case class MultilineString(userConfig: () => UserConfiguration)
...
   }
 
+  private def wasTripleQuoted(
+      splitLines: Array[String],
+      position: Position,
+  ): Boolean = {
+    val currentLine = splitLines(position.getLine - 1)
+    getIndexOfLastTripleQuote(currentLine).exists { case (_, quoteClosed) =>
+      !quoteClosed
+    }
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:276]:
```suggestion
    val prevLine = splitLines(position.getLine - 1)
    val currLine = splitLines(position.getLine)
    prevLine.endsWith("\"\"\"") && !currLine.startsWith("\"\"\"")
```

Can we simplify it like this?
The idea is, that what we need to know are:
- Did the user enter a newline after `"""`?
- Is the triple quote closed (that means the current line of cursor starts with `"""`)? (maybe there's some whitespace for indent)

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala [hunk 2]
```diff
@@ -100,9 +134,9 @@ case class MultilineString(userConfig: () => UserConfiguration)
   }
 
   private def hasNQuotes(start: Int, text: String, n: Int): Boolean =
-    (start until start + n).forall(i =>
+    (start until start + n).forall { i =>
       if (i < 0 || i > text.length) false else text(i) == quote
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:138]: Not related to this change, but I encounter some `IndexOutofBoundsException` when I write quote in the end of a file, maybe we should check the length of text before accessing `text(I)` :thinking:

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala [hunk 3]
```diff
@@ -276,6 +320,12 @@ case class MultilineString(userConfig: () => UserConfiguration)
     List(new TextEdit(new Range(pos1, pos2), "\"\"\"\"\"\""))
   }
 
+  private def addTripleQuote(pos: Position): List[TextEdit] = {
+    val pos1 = new Position(pos.getLine, pos.getCharacter)
+    val pos2 = new Position(pos.getLine, pos.getCharacter + 1)
+    List(new TextEdit(new Range(pos1, pos2), "\"\"\""))
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:326]: [nit]
```suggestion
    List(new TextEdit(new Range(pos, pos), "\"\"\""))
```

It's ok with point

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 1 above)
Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:276]: Thank you so much for reviewing this Rikito :)

Well, frankly this was copied from `doubleQuoteNotClosed` whose logic was previously tested in the code.

As an example of the issue's trickiness, the `getIndexOfLastTripleQuote`, which was copied from `getIndexOfLastQuote` does not just get the index of last triple quote, but also It gets the index of the the last triple quote, which was not closed. 

So basically this prevents us from putting an extra triple quote if the previous line had two triple quotes: one begin, and one close.

Another tricky matter is that, this code action does itself generate a text edit to put the closing triple quotes in the `curLine`which is the line after the newly inserted newline character. 

So there won't be any triple quotes in it.

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 2 above)
Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:138]: hmmm, it seems that it is checking `i > text.length`

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 3 above)
Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:326]: I am just afraid about removing the `pos.getCharacter + 1` for the end of range, because this was copied from existing logic and the indices in TextEdits are so tricky.

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 2 above)
Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala]: Ah, I got it, if i == text.length it will fail

Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala]: it should be i >= text.length

Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala]: I wonder why it is if `(i < 0 || i > text.length)` in the first place, it is not checking the `0` index either.

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 1 above)
Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:276]: Nevermind, @vzmerr pointed me out that it doesn't work for `"""foo|"""` and enter newline.

Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:276]: I just noticed that the suggestions might lead to unexpected results in a few cases, for example
`val a = """" this is an open triple quoted string not ending with triple quotes`

Also, the function `getIndexOfLastTripleQuote` does more than just escaping the escaped ones. It also keeps opening and closing the triple quotes of the previous line, as it traverses through the characters to make sure that the last triple quote in the line is still open.
`a = """ first one """ + """ second one """ + """ third one which is not closed`

Also in case of `sql"""` the closing triple quotes do not get inserted automatically, so they are not present in the first place.

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala [hunk 4]
```diff
@@ -87,10 +88,44 @@ case class MultilineString(userConfig: () => UserConfiguration)
...
+            lastTripleQuote = i
+            tripleQuoteClosed = !tripleQuoteClosed
+            quoteNum = 0
+          }
+        } else {
+          escaped = !escaped
+        }
+      } else if (char == '\\') {
+        quoteNum = 0
+        escaped = !escaped
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: I still don't think escape handling is not needed because of https://github.com/scalameta/metals/pull/4151#discussion_r921969587

**Review by tanishiking — COMMENTED** [review]
I left one comment, but otherwise LGTM!

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 4 above)
Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: we are traversing character by character. What if the escape is at the beginning?

Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: `\"""`

Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: ` val a = """ This is a sample \""" quoted string in which escaping matters

Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: Ah, right

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala [hunk 5]
```diff
@@ -227,11 +262,30 @@ case class MultilineString(userConfig: () => UserConfiguration)
...
     }
   }
 
+  private def wasTripleQuoted(
+      splitLines: Array[String],
+      position: Position,
+  ): Boolean = {
+    var closedFromPreviousLines = true
+    var existed = false
+    for (i <- 0 until position.getLine()) {
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:276]: It's a pity we need to go over every line, but I don't see a way around it :thinking:

**Review by tgodzik — APPROVED** [review]
LGTM!

**Review by tanishiking — COMMENTED** [review]
LGTM!
but can you add those cases to test? it's ok in another PR :)

**Comment by vzmerr** [comment]
> LGTM!
> but can you add those cases to test? it's ok in another PR :)
> [#4151 (comment)](https://github.com/scalameta/metals/pull/4151#discussion_r921987100)
> [#4151 (comment)](https://github.com/scalameta/metals/pull/4151#discussion_r923176489)

good idea 👍

File: metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala (hunk 4 above)
Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: turns out that you were right @tanishiking 

In multiline strings, one cannot escape `"` using `\` character.

Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: https://scastie.scala-lang.org/vzmerr/tPvC34B4RM2EIywhJlcLRg/46

Comment (vzmerr) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: https://scastie.scala-lang.org/vzmerr/tPvC34B4RM2EIywhJlcLRg/47

Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/MultilineString.scala:114]: Thank you for investigating that!
